pkg/manager/impl/dataset_manager.go

```diff
@@ -129,6 +132,58 @@ func (dm *datasetManager) GetDataset(ctx context.Context, request datacatalog.Ge
 	}, nil
 }
 
+// List Datasets with optional filtering and pagination
+func (dm *datasetManager) ListDatasets(ctx context.Context, request datacatalog.ListDatasetsRequest) (*datacatalog.ListDatasetsResponse, error) {
+	err := validators.ValidateListDatasetsRequest(&request)
+	if err != nil {
+		logger.Warningf(ctx, "Invalid list datasets request %v, err: %v", request, err)
+		dm.systemMetrics.validationErrorCounter.Inc(ctx)
+		return nil, err
+	}
+
+	// Get the list inputs
+	listInput, err := transformers.FilterToListInput(ctx, common.Dataset, request.GetFilter())
+	if err != nil {
+		logger.Warningf(ctx, "Invalid list datasets request %v, err: %v", request, err)
+		dm.systemMetrics.validationErrorCounter.Inc(ctx)
+		return nil, err
+	}
+
+	err = transformers.ApplyPagination(request.Pagination, &listInput)
+	if err != nil {
+		logger.Warningf(ctx, "Invalid pagination options in list datasets request %v, err: %v", request, err)
+		dm.systemMetrics.validationErrorCounter.Inc(ctx)
+		return nil, err
+	}
+
+	// Perform the list with the dataset and listInput filters
+	datasetModels, err := dm.repo.DatasetRepo().List(ctx, listInput)
+	if err != nil {
+		logger.Errorf(ctx, "Unable to list Datasets err: %v", err)
+		dm.systemMetrics.listFailureCounter.Inc(ctx)
+		return nil, err
+	}
+
+	// convert returned models into entity list
+	datasetList := make([]*datacatalog.Dataset, len(datasetModels))
+	for idx, datasetModel := range datasetModels {
+		dataset, err := transformers.FromDatasetModel(datasetModel)
+		if err != nil {
+			logger.Errorf(ctx, "Unable to transform Dataset %+v err: %v", datasetModel, err)
```

Comment:
how big are datasetModels? maybe just log the DatasetKey here for brevity?

Reply:
that's a good idea since the metadata can be large